% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{article}
\title{Vignette1}
\author{Dietrich Epp Schmidt}
\date{11/30/2021}

\usepackage{amsmath,amssymb}
\usepackage{lmodern}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\hypersetup{
  pdftitle={Vignette1},
  pdfauthor={Dietrich Epp Schmidt},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}
\urlstyle{same} % disable monospaced font for URLs
\usepackage[margin=1in]{geometry}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{-\maxdimen} % remove section numbering
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi

\begin{document}
\maketitle

\hypertarget{intro}{%
\subsection{INTRO}\label{intro}}

This package is meant to facilitate downloading and summarizing the
annotations of completed GenBank and RefSeq prokaryote genome
assemblies. It also contains functionality to add gene annotations to a
phyloseq object's taxonomy table; or to subset a phyloseq object to only
taxa that contain a set of genes. This tool is not meant to approximate
or infer metagenomes, though it operates on many of the same principles
as other tools that do. Also, this was built to work on a macOS
platform. Windows and Linux are not supported. If you want, you may fork
the repository and add that functionality.

To install this package:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{install.packages}\NormalTok{(}\StringTok{"devtools"}\NormalTok{)}
\FunctionTok{library}\NormalTok{(devtools)}
\NormalTok{devtools}\SpecialCharTok{::}\FunctionTok{install\_github}\NormalTok{(}\StringTok{\textquotesingle{}Djeppschmidt/ProkaryoteGeneAnnotation\textquotesingle{}}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

There are three steps to this process: 1) get the reference data -
feature tables from RefSeq and GenBank 2) summarize the tables by genus,
species, and strain identifiers 3) subset, or modify phyloseq objects
based on these identifiers

\hypertarget{get-reference-data}{%
\subsection{Get reference data}\label{get-reference-data}}

The first step is to download the most up to date reference data. NCBI
has a functionality to make a table that contains accession information
for any data. We start by using their tool to filter to only prokaryote
taxa, and remove any assemblies that are not complete. This can be done
by following this link:

\url{https://www.ncbi.nlm.nih.gov/genome/browse\#!/prokaryotes/}

Click the filters dropdown menu on the upper right hand side of the data
table. Next to the Assembly level tab, check the box for ``complete''.
Then click the download icon at the top left of the data table (below
the filters).

Name this file whatever you want, and put it in whatever directory you
want. In this example, it is named ``prokaryote.csv''

Import this file into R:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# import reference dataset:}
\CommentTok{\# refdir from https://www.ncbi.nlm.nih.gov/genome/browse\#!/prokaryotes/}
\CommentTok{\# filter set to only bacteria and archaea}
\CommentTok{\# and only whole genomes}
\NormalTok{refdir}\OtherTok{\textless{}{-}}\FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{as.matrix}\NormalTok{(}\FunctionTok{read.csv}\NormalTok{(}\StringTok{"\textasciitilde{}/Documents/GitHub/SoilHealthDB/prokaryotes.csv"}\NormalTok{, }\AttributeTok{sep=}\StringTok{","}\NormalTok{)))}
\end{Highlighting}
\end{Shaded}

Next we need to download the annotated feature tables. These tables are
placed in an output directory of your choice, in a folder corresponding
to the source of the annotations (GenBank or RefSeq). There are several
inputs required for this function:

\begin{enumerate}
\def\labelenumi{\arabic{enumi})}
\tightlist
\item
  the reference data frame made in the first step (refdir)
\item
  an output directory path (outpath)
\item
  a directory path to the bin folder where wget is compiled (you may
  need to install this)
\end{enumerate}

The bin folder where wget is compiled may not be in the same location
bin folder that RStudio references. This function assumes wget is
compiled in ``/usr/local/bin/''. If this is not the case on your
machine, change the binPATH option.

Finally, make sure that your output directory has a large enough volume
to accommodate at least 30 GB of data. I sometimes use an external hard
drive for this type of task.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# run:}
\NormalTok{outpath}\OtherTok{\textless{}{-}}\StringTok{"/Downloads"}
\NormalTok{binPath}\OtherTok{\textless{}{-}}\StringTok{"/usr/local/bin/"}
\FunctionTok{download.Feature.Tables}\NormalTok{(refdir, outpath, binPATH)}
\end{Highlighting}
\end{Shaded}

\hypertarget{summarize-the-reference-tables}{%
\subsection{Summarize the reference
tables}\label{summarize-the-reference-tables}}

This function takes a vector of gene names / symbols. An example is
``nifH''. You may include as many as you like in this vector. The output
will be a table with Genus, Species, Strain, Accession number, and a
column for each gene. Rows are unique genome assemblies. The value in
the column will be zero if the assembly does not have the gene; and 1 if
it does.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# run:}
\CommentTok{\# refdir = same reference table from above}
\CommentTok{\# directory = same directory from the previous function (where the GenBank and RefSeq folders exist)}
\CommentTok{\# genes = vector of genes}
\NormalTok{gene.tab}\OtherTok{\textless{}{-}}\FunctionTok{get.genes}\NormalTok{(refdir, outpath, genes)}
\end{Highlighting}
\end{Shaded}

\hypertarget{annotate-or-manipulate-phyloseq-objects-using-this-reference-data}{%
\subsection{Annotate or manipulate phyloseq objects using this reference
data}\label{annotate-or-manipulate-phyloseq-objects-using-this-reference-data}}

This package offers the ability to subset phyloseq objects to isolate
the taxa that likely have certain genes. This works by subsetting the
phyloseq taxonomy based on matches to either the species or genus level.
I strongly recommend using the species level cutoff.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{subset.phyloseq}\NormalTok{(phyloseq, genes, level)}
\end{Highlighting}
\end{Shaded}

And that's it.

\end{document}
